pkg/util/modelutil/merger.go

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cortexproject/cortex/blob/master/pkg/util/merger.go
+// Provenance-includes-location: https://github.com/thanos-io/thanos/blob/main/pkg/strutil/merge.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Cortex Authors.
+// Provenance-includes-copyright: The Thanos Authors.
+
+package modelutil
+
+import (
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+// MergeSampleSets merges and dedupes two sets of already sorted sample pairs.
+func MergeSampleSets(a, b []model.SamplePair) []model.SamplePair {
+	result := make([]model.SamplePair, 0, len(a)+len(b))
+	i, j := 0, 0
+	for i < len(a) && j < len(b) {
+		if a[i].Timestamp < b[j].Timestamp {
+			result = append(result, a[i])
+			i++
+		} else if a[i].Timestamp > b[j].Timestamp {
+			result = append(result, b[j])
+			j++
+		} else {
+			result = append(result, a[i])
+			i++
+			j++
+		}
+	}
+	// Add the rest of a or b. One of them is empty now.
+	result = append(result, a[i:]...)
+	result = append(result, b[j:]...)
+	return result
+}
+
+// MergeNSampleSets merges and dedupes n sets of already sorted sample pairs.
+func MergeNSampleSets(sampleSets ...[]model.SamplePair) []model.SamplePair {
+	l := len(sampleSets)
+	switch l {
+	case 0:
+		return []model.SamplePair{}
+	case 1:
+		return sampleSets[0]
+	}
+
+	n := l / 2
+	left := MergeNSampleSets(sampleSets[:n]...)
+	right := MergeNSampleSets(sampleSets[n:]...)
+	return MergeSampleSets(left, right)
+}
+
+// MergeHistogramSets merges and dedupes two sets of already sorted histograms.
+func MergeHistogramSets(a, b []mimirpb.Histogram) []mimirpb.Histogram {
```

Comment:
There is a potential to use generics here. I am not sure if you have generics in Mimir or want to introduce it. Do you mind if I give it a try?

Reply:
I'm on the fence about generics atm since it adds some overhead in the floats case. I'd rather we do that in a separate PR with benchmarks

Reply:
> I'm on the fence about generics atm since it adds some overhead in the floats case

Are you sure you are not confusing generics with the use of `interface{}`?

Reply:
From offline discussion, turns out that if you use methods with generics, they still use the same mechanism as the regular interface. So it could indeed be slower.